### PR TITLE
Added Adam Optimizer

### PR DESCRIPTION
**Pull request by jrego43**
Added `Adam` struct that implements the `Optimizer` trait, closely following its PyTorch implementation. It essentially combines the two types of momentums found in `SGD` and `RMSProp` (with respective decay rates β1 and β2) by keeping track of several gradient-based averages.

Source: https://pytorch.org/docs/stable/generated/torch.optim.Adam.html#torch.optim.Adam

### CODE REVIEW

File: src/optimizers/adam.rs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+                .unwrap_or_else(|| panic!("Optimizer cannot step when gradient is None."))
+                .value();
+
+            if self.maximize {
+                grad = -grad;
+            }
+            grad += param.value() * self.weight_decay;
+
+            exp_avgs[i] = (beta1 * exp_avgs[i]) + ((1. - beta1) * grad);
+            exp_avg_sqs[i] = (beta2 * exp_avg_sqs[i]) + ((1. - beta2) * grad.pow(2));
```
Comment (LazaroHurtado) [inline src/optimizers/adam.rs:67]: I believe you had to `use num_traits::Pow` because of the `grad.pow(2)` but you can do without the import by doing `grad.powf(2.0)`

**Review by LazaroHurtado — CHANGES_REQUESTED** [review]
LGTM, just a small question about an import

File: src/optimizers/adam.rs (hunk 1 above)
Comment (jrego43) [reply src/optimizers/adam.rs:67]: You're absolutely right, but the reason I did it is because `RMSProp` does `use num_traits::Pow` as well. I can go ahead and make the change to both, though!

**Review by LazaroHurtado — APPROVED** [review]
Awesome job!

**Review by Demba3 — APPROVED** [review]
